File: androidApp/app/src/main/java/us/mrvivacio/porno/Utilities.java
package us.mrvivacio.porno;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Environment;
import android.util.Log;

import androidx.preference.PreferenceManager;

import java.io.File;
import java.util.ArrayList;
import java.util.Map;

public class Utilities {
    public static final String LINKS_MIGRATED = "migration_done_v1";
    private static final String PREFIX = "url_"; // Our new namespace

    /**
     * Migrates data from the legacy Text File and Activity-specific SharedPreferences
     * into the Global DefaultSharedPreferences.
     */
    public static void migrateIfNecessary(Context context) {
        SharedPreferences globalPrefs = PreferenceManager.getDefaultSharedPreferences(context);
        if (globalPrefs.getBoolean(LINKS_MIGRATED, false)) {
            return;
        }

        SharedPreferences.Editor editor = globalPrefs.edit();

        // Migrate from OLD Activity SharedPreferences (Preserves Name -> URL mapping)
        // Trivia: previously, used PorNo_user_data/links.txt
        SharedPreferences oldPrefs = context.getSharedPreferences("MainActivity", Context.MODE_PRIVATE);
        Map<String, ?> allEntries = oldPrefs.getAll();
        for (Map.Entry<String, ?> entry : allEntries.entrySet()) {
            if (entry.getValue() instanceof String) {
                Log.d("PorNo_migration", (entry.getValue()).toString());

                // ADD PREFIX DURING MIGRATION
                editor.putString(PREFIX + entry.getKey(), (String) entry.getValue());
            }
        }

        deleteLegacyFolder(context);
        oldPrefs.edit().clear().apply();

        editor.putBoolean(LINKS_MIGRATED, true);
        editor.apply();
    }

    /**
     * Returns a random URL from the global SharedPreferences.
     * Used by MyAccessibilityService.
     */
    public static String getRandomURL(Context context) {
        SharedPreferences globalPrefs = PreferenceManager.getDefaultSharedPreferences(context);
        Map<String, ?> allEntries = globalPrefs.getAll();

        ArrayList<String> urlsOnly = new ArrayList<>();

        for (Map.Entry<String, ?> entry : allEntries.entrySet()) {
            if (entry.getKey().startsWith(PREFIX) && entry.getValue() instanceof String) { // skips LINKS_MIGRATED flag
                urlsOnly.add((String) entry.getValue());
            }
        }

        if (!urlsOnly.isEmpty()) {
            int random = (int) (Math.random() * urlsOnly.size());
            return urlsOnly.get(random);
        }

        // Fallback if no links are found
        return "https://medium.com/@vivekbhookya/porno-de97189d82f6";
    }

    /**
     * Standard save method for the new SharedPreferences architecture
     */
    public static void saveUrl(Context context, String name, String url) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        // ADD PREFIX ON SAVE
        prefs.edit().putString(PREFIX + name, url).apply();
    }

    /**
     * Standard remove method for the new SharedPreferences architecture
     */
    public static void removeUrl(Context context, String name) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(PREFIX + name).apply();
    }

    private static void deleteLegacyFolder(Context context) {
        try {
            // Locations to check
            File[] roots = {
                    Environment.getExternalStorageDirectory(), // The "Internal Storage" root
                    context.getExternalFilesDir(null),         // App-specific external
                    new File("/sdcard")                        // Common alias on older Samsungs
            };

            for (File root : roots) {
                if (root == null) continue;

                // Look for the folder in this root
                File dir = new File(root, "PorNo_user_data");

                if (dir.exists() && dir.isDirectory()) {
                    // 1. Delete the file inside first
                    File file = new File(dir, "links.txt");
                    if (file.exists()) {
                        boolean fileDeleted = file.delete();
                        Log.d("PorNo_Cleanup", "File deleted: " + fileDeleted);
                    }

                    // 2. Delete the folder now that it's empty
                    boolean dirDeleted = dir.delete();
                    Log.d("PorNo_Cleanup", "Folder deleted: " + dirDeleted + " at " + dir.getAbsolutePath());
                }
            }
        } catch (Exception e) {
            Log.e("PorNo_Cleanup", "Failed cleanup: " + e.getMessage());
        }
    }
}

File: androidApp/app/src/main/java/us/mrvivacio/porno/Utilities.kt
package us.mrvivacio.porno

import android.os.Environment
import android.util.Log
import java.io.BufferedReader
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.FileReader
import java.io.IOException
import kotlin.math.floor

val randomURL: String
    get() {
        val reader = getReader("links.txt") ?: return "https://medium.com/@vivekbhookya/porno-de97189d82f6"

        try {
            var currLine = reader.readLine()
            val currLinks = mutableListOf<String>()

            // While we still have text to read, add all the links
            while (currLine != null) {
                // Add the links
                currLinks.add(currLine)
                currLine = reader.readLine()
            }

            // Now choose a link at random
            return if (currLinks.isEmpty()) {
                "https://medium.com/@vivekbhookya/porno-de97189d82f6"
            } else {
                val random = floor(Math.random() * currLinks.size).toInt()
                currLinks[random]
            }
        } catch (ioe: IOException) {
            Log.e("Utilities", "Error reading links file", ioe)
            return "https://medium.com/@vivekbhookya/porno-de97189d82f6"
        } finally {
            reader.close()
        }
    }


fun getDirectory(): File? {
    val path = Environment.getExternalStorageDirectory()
    val dir = File("${path.absolutePath}/PorNo_user_data/")
    if (!dir.isDirectory && !dir.mkdir()) {
        // Directory creation failed
        Log.e("Utilities", "Failed to create directory: ${dir.absolutePath}")
        return null
    }

    return dir
}

fun getFile(fileName: String): File? {
    val dir = getDirectory() ?: return null // Return null if getDirectory fails
    val file = File(dir, fileName)
    if (!file.exists()) {
        try {
            file.createNewFile()
        } catch (e: IOException) {
            Log.e("Utilities", "Error creating file: ${file.absolutePath}", e)
            return null
        }
    }

    return file
}

fun getReader(fileName: String): BufferedReader? {
    val file = getFile(fileName) ?: return null
    try {
        return BufferedReader(FileReader(file))
    } catch (e: FileNotFoundException) {
        Log.e("Utilities", "File not found: ${file.absolutePath}", e)
        return null
    }
}

fun getWriter(fileName: String): FileOutputStream? {
    val file = getFile(fileName) ?: return null
    try {
        return FileOutputStream(file)
    } catch (e: FileNotFoundException) {
        Log.e("Utilities", "File not found: ${file.absolutePath}", e)
        return null
    }
}

fun saveToFile(URLs: List<String>) {
    val os = getWriter("links.txt") ?: return

    var urls = ""
    for (url in URLs) {
        urls += "$url\n"
    }

    try {
        os.write(urls.toByteArray())
        Log.d("Utilities", "CREATING NEW FILE")

        os.close()
    } catch (e: IOException) {
        Log.e("Utilities", "Error writing to file", e)
    }
}

fun updateFile(url: String) {
    val file = getFile("links.txt") ?: return // If getFile() returns null, exit
    val urlWithNewLine = "$url\n"

    // Check if the file exists. If not, write the url to a new file
    if (!file.exists()) {
        val os = getWriter("links.txt") ?: return
        try {
            os.write(urlWithNewLine.toByteArray())
            os.close()
        } catch (e: IOException) {
            Log.e("Utilities", "Error writing to new file", e)
        }
    } else {
        try {
            // Open in append mode
            val os = FileOutputStream(file, true)
            os.write(urlWithNewLine.toByteArray())
            os.close()
        } catch (ioe: IOException) {
            Log.e("Utilities", "Error appending to file", ioe)
        }
    }
}

fun removeFromFile(url: String) {
    val file = getFile("links.txt") ?: return

    if (!file.exists()) {
        // no file no work
        return
    } else {
        val reader = getReader("links.txt") ?: return

        try {
            var currLine = reader.readLine()
            var currLinks = ""

            // While we still have text to read, add all the links
            while (currLine != null) {
                // Add the links but the one to delete
                if (!currLine.contains(url)) {
                    currLinks += "$currLine\n"
                }

                currLine = reader.readLine()
            }

            // Now save all the links
            val os = FileOutputStream(file)

            os.write(currLinks.toByteArray())

            // Our job here is done
            os.close()
        } catch (ioe: IOException) {
            Log.e("Utilities", "Error removing URL from file", ioe)
        } finally {
            reader.close()
        }
    }
}